@rules endpoint #1411 (#1496)
plone/plone.restapi: docs/source/endpoints/rules.md

---
html_meta:
  "description": "Available content rules in a Plone site can be listed and queried by accessing the /@content-rules endpoint on any context. Access requires the ManageContentRules permission."
  "property=og:description": "Available content rules in a Plone site can be listed and queried by accessing the /@content-rules endpoint on any context. Access requires the ManageContentRules permission."
  "property=og:title": "Content rules"
  "keywords": "Plone, plone.restapi, REST, API, ContentRules, Rules"
---

(rules)=

# Content rules

```{note}
These docs are generated by code tests, therefore you will see some `test` rules appear here.
```

Available content rules in a Plone site can be listed and queried by accessing the `/@content-rules` endpoint on any context.
Access requires the ManageContentRules permission.

```{eval-rst}
..  http:example:: curl httpie python-requests
    :request: ../../src/plone/restapi/tests/http-examples/rules_get.req
```

```{literalinclude} ../../src/plone/restapi/tests/http-examples/rules_get.resp
:language: http
```

The API consumer can assign, unassign , enable, disable,

> (disable) apply to subfolders any of the rules available in the portal.

Reviewer: Can you better explain what is intent of "(disable) apply"? Does it mean the "opposite of apply"? If so, then it should be "remove, and apply".

| Verb    | URL              | Action                                             |
| ------- | ---------------- | -------------------------------------------------- |
| `POST`  | `/@content-rules/{rule-id}`| Add rule to context                                          |
| `GET`   | `/@content-rules/`         | Get aquired, assignable and assigned rules                      |

> | `PATCH` | `/@content-rules/` with RAW Body | enable/disable/(disable) apply to subfolders/move down/up |

Reviewer: See previous note about "(disable) apply".

Also what does "RAW Body" mean? Why is it uppercase?

| `DELETE`| `/@content-rules/` with RAW Body | Unassign rule on context                                  |

## Assigning a new Content rule with `POST`

To assign a content rule to a context, send a `POST` request to the `<context>/@content-rules` endpoint:

```{eval-rst}
..  http:example:: curl httpie python-requests
    :request: ../../../src/plone/restapi/tests/http-examples/rules_add.req
```

Response:

```{literalinclude} ../../../src/plone/restapi/tests/http-examples/rules_add.resp
:language: http
```

> ## Getting content rule on a context with `GET`

Reviewer: ```suggestion
## Getting content rule for a context with `GET`
```

> To get content rules assigned/assignable/aquired on a context, send a `GET`

Reviewer: ```suggestion
To get content rules that are assigned, assignable, or acquired for a context, send a `GET`
```

request to the `<context>/@content-rules` endpoint:

```{eval-rst}
..  http:example:: curl httpie python-requests
    :request: ../../../src/plone/restapi/tests/http-examples/rules_get.req
```

Response:

```{literalinclude} ../../../src/plone/restapi/tests/http-examples/rules_get.resp
:language: http
```

> ## Changing content rules on a context with `PATCH`

Reviewer: ```suggestion
## Changing content rules for a context with `PATCH`
```

> To make changes on content rule assignments on a context, send a `PATCH`

Reviewer: ```suggestion
To make changes on content rule assignments for a context, send a `PATCH`
```

request to the `<context>/@content-rules` endpoint:

### Apply on subfolders

```{eval-rst}
..  http:example:: curl httpie python-requests
    :request: ../../../src/plone/restapi/tests/http-examples/rules_apply_subfolders.req
```

Response:

```{literalinclude} ../../../src/plone/restapi/tests/http-examples/rules_apply_subfolders.resp
:language: http
```

> ### Disable apply on subfolders

Reviewer: The request and response are identical to those from the previous subsection. Maybe I don't know the difference between "Apply" and "Disable Apply", but I would expect the examples to reflect whatever that difference may be.

Author: This was a very good catch! Should be fixed now.

Reviewer: Yup, looks good.

---
myst:
  html_meta:
  "description": "Available content rules in a Plone site can be listed and queried by accessing the /@content-rules endpoint on any context. Access requires the ManageContentRules permission."
  "property=og:description": "Available content rules in a Plone site can be listed and queried by accessing the /@content-rules endpoint on any context. Access requires the ManageContentRules permission."
  "property=og:title": "Content Rules"

>   "keywords": "Plone, plone.restapi, REST, API, ContentRules, Rules"

Reviewer: The above 4 keys ("description", etc.) need 2 more spaces of indentation.

---

(content-rules)=

# Content Rules

```{note}
These docs are generated by code tests, therefore you will see some `test` rules appear here.
```

Available content rules in a Plone site can be listed and queried by accessing the `/@content-rules` endpoint in any context.
Access requires the `ManageContentRules` permission.

```{eval-rst}
..  http:example:: curl httpie python-requests
    :request: ../../../src/plone/restapi/tests/http-examples/rules_get.req

> ```

Reviewer: The tabs exploded throughout the page. This might be due to the lack of indentation for the `html_meta` keys.

![Screen Shot 2022-09-28 at 12 36 03 PM](https://user-images.githubusercontent.com/102112/192872921-2ce7a65c-b5ee-464f-9397-565958416208.png)

Reviewer: These still do not display correctly in the Preview on Netlify. I am not sure why.

I pulled down this PR to troubleshoot, and found two other bugs that I fixed in two other commits. When I do a `make docs-livehtml`, then these display just fine. `¯\_(ツ)_/¯`. I'm going to resolve this conversation.
